blog/_posts/2023-12-08-migration-to-scala-3.md

```diff
@@ -0,0 +1,146 @@
+---
+layout: blog-detail
+post-type: blog
+by: Zelenya
+title: "Scala 3: My migration journey"
+description: I want to share my journey from Scala 2 to Scala 3 and discuss migration to Scala 3
+---
+
+## Implicit context
+
+Before doing any actual personal work, based on the internet discussions, I had an impression that switching to Scala 3 would be quite painful — in practice, it was pretty straightforward.
+
+In the retrospect, this was the case because I (and my team) fall into the following buckets:
+
+- Don’t use IntelliJ;
+- Don’t use macros;
+- Use Typelevel stack.
+
+It wouldn’t be as smooth if any of these were different.
+
+Let’s start in my world first, follow the primary migration steps, and then talk about what to do if your setup and stack differ.
+
+## Step 0. What does “migration” mean in your case?
+
+Before you can jump into the action, you (and your team) have to answer a couple of questions.
+
+If you are actively writing Scala, I’d expect that you don’t choose to stay on Scala 2. Why would anyone want to miss out on all the new features and improvements? Who wouldn’t want to rewrite everything to Scala 3? The reality is not so simple. Some people have monoliths, some use Spark ([which you can already run with Scala 3](https://xebia.com/blog/using-scala-3-with-spark/)), some have no time, and so on.
+
+There is no one way to do the migration. Decide what’s possible and worthwhile to you (and your company).
+
+Should you rewrite one service at a time? In what order? Or should you leave the old services untouched and write all the new services in Scala 3? What about internal libraries? How much time and resources can you allocate? (The last one might be the hardest to address)
```

Comment:
Here you're making an implicit assumption that the users work with the microservices architecture, and proceed to explore the decision space of how to migrate such an architecture. Later on, you mention your decision on how to do it.

As you also mention, different people might have different architectures. It would be, therefore, good to spell out explicitly that these questions apply to a particular architecture of your choice.

E.g. if a person is a library author, the questions about in which order to port services or whether to port all of them might not make much sense to them.

Reply:
It would actually be good to put this information about the architecture in the introductory section I mentioned in another comment, since it seems this architectural assumption also drives the rest of the article.